src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -43,32 +132,51 @@ private abstract class TransferOperation : AsyncOperation
             public int ReceivedFlags;
         }
 
-        private sealed class SendReceiveOperation : TransferOperation
+        private abstract class SendReceiveOperation : TransferOperation
         {
-            public Action<int, byte[], int, int, SocketError> Callback;
+            public Action<int, byte[], int, int, SocketError> Callback { set { CallbackOrEvent = value; } }
             public BufferList Buffers;
             public int BufferIndex;
 
-            public override void Complete()
+            protected sealed override void DoCallback()
             {
-                Debug.Assert(Callback != null);
+                var callback = (Action<int, byte[], int, int, SocketError>)CallbackOrEvent;
```

Comment:
Similarly, might make sense to encapsulate this in the property.  That way the call sites don't need to understand how/where it's actually stored.